routing/triert/table.go

```diff
@@ -0,0 +1,227 @@
+package triert
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	tkey "github.com/libp2p/go-libp2p-xor/key"
+	"github.com/libp2p/go-libp2p-xor/trie"
+	"github.com/plprobelab/go-kademlia/key"
+	"github.com/plprobelab/go-kademlia/network/address"
+)
+
+// TODO: move to key package and replace ErrInvalidKey
+// ErrKeyWrongLength indicates a key has the wrong length
+var ErrKeyWrongLength = errors.New("key has wrong length")
+
+// TrieRT is a routing table backed by a Xor Trie which offers good scalablity and performance
+// for large networks. All exported methods are safe for concurrent use.
+type TrieRT struct {
+	self key.KadKey
+
+	// keymu is held during keys and keyNodes mutations to serialize changes
+	keymu sync.Mutex
+
+	// keys holds a pointer to an immutable trie
+	// any store to keys must be performed while holding keymu, loads may be performed without the lock
+	keys atomic.Value
+
+	// keyNodes holds a mapping of kademlia key to node id (map[string]address.NodeID)
+	// note: this could be eliminated if the trie was modified to allow a tuple to be stored at each node
+	// any store to keyNodes must be performed while holding keymu, loads may be performed without the lock
+	keyNodes atomic.Value
+}
+
+func New(self key.KadKey) *TrieRT {
```

Comment:
This routing table is unlimited in size. It would be great to limit the size of each bucket, if we go for the _normal server_ routing table implementation.

Reply:
I think it would be useful to have full control over the maximum population of each bucket. We could actually implement this as a filter. Allow a function to be specified with the following signature that is called by the rt in AddPeer:

```Go
func(rt *TrieRT, kk key.KadKey) bool
```

Then we can pass a filter that limits buckets to 20
```Go
func Bucket20(rt *TrieRT, kk key.KadKey) bool {
    cpl := rt.Cpl(kk)  // this doesn't exist but it basically does rt.self.CommonPrefixLength(kk)
    return rt.CplSize(cpl) < 20
}
```

Or one that allows unlimited peers: 

```Go
func Unlimited(rt *TrieRT, kk key.KadKey) bool {
    return true
}
```

Or one that varies occupancy by cpl:

```Go
func VaryOccupancy(rt *TrieRT, kk key.KadKey) bool {
    cpl := rt.Cpl(kk)
    return rt.CplSize(cpl) < 5*cpl
}
```

Reply:
Done